### DISCUSSION TITLE
phase locked loop for framerate

### DISCUSSION THREAD

**Discussion by michaelchimento** [raspberrypi/picamera2#553] [Q&A]
hi, i'm interested in implementing a PLL to adjust the framerate for the v3 camera (similar to a previous implementation in raspivid https://github.com/inastitch/raspivid-inatech), although tbh libcamera is greek to me, and i'd rather write it in python. I figured I might save myself some time if someone could point me to where the frame callbacks happen in picamera2. Thanks for your help!

**Comment by davidplowman** [comment]
Hi, you can set the framerate directly as follows:
```
picam2.set_controls({'FrameRate': 30.0})
```
Does that help?
